performance/shared/src/main/kotlin/org/jetbrains/benchmarksLauncher/launcher.kt

```diff
@@ -73,36 +146,28 @@ abstract class Launcher {
         val runningBenchmarks = if (filterSet.isNotEmpty() || regexes.isNotEmpty()) {
             benchmarks.filterKeys { benchmark -> benchmark in filterSet || regexes.any { it.matches(benchmark) } }
         } else benchmarks
-        if (runningBenchmarks.isEmpty())
+        if (runningBenchmarks.isEmpty()) {
+            printStderr("No matching benchmarks found\n")
             error("No matching benchmarks found")
+        }
         val benchmarkResults = mutableListOf<BenchmarkResult>()
         for ((name, benchmark) in runningBenchmarks) {
-            val benchmarkInstance = (benchmark as? BenchmarkEntryWithInit)?.ctor?.invoke()
-            var i = numWarmIterations
-            logger.log("Warm up iterations for benchmark $name\n")
-            runBenchmark(benchmarkInstance, benchmark, i)
-            var autoEvaluatedNumberOfMeasureIteration = 1
-            while (true) {
-                var j = autoEvaluatedNumberOfMeasureIteration
-                val time = runBenchmark(benchmarkInstance, benchmark, j)
-                if (time >= 100L * 1_000_000) // 100ms
-                    break
-                autoEvaluatedNumberOfMeasureIteration *= 2
+            val recordMeasurement : (RecordTimeMeasurement) -> Unit = {
+                benchmarkResults.add(BenchmarkResult(
+                    "$prefix$name",
+                    it.status,
+                    it.durationNs / 1000,
+                    BenchmarkResult.Metric.EXECUTION_TIME,
+                    it.durationNs / 1000,
+                    it.iteration + 1,
+                    it.warmupCount))
             }
-            logger.log("Running benchmark $name ")
-            val samples = DoubleArray(numberOfAttempts)
-            for (k in samples.indices) {
-                logger.log(".", usePrefix = false)
-                i = autoEvaluatedNumberOfMeasureIteration
-                val time = runBenchmark(benchmarkInstance, benchmark, i)
-                val scaledTime = time * 1.0 / autoEvaluatedNumberOfMeasureIteration
-                samples[k] = scaledTime
-                // Save benchmark object
-                benchmarkResults.add(BenchmarkResult("$prefix$name", BenchmarkResult.Status.PASSED,
-                        scaledTime / 1000, BenchmarkResult.Metric.EXECUTION_TIME, scaledTime / 1000,
-                        k + 1, numWarmIterations))
+            try {
+                runBenchmark(logger, numWarmIterations, numberOfAttempts, name, recordMeasurement, benchmark)
+            } catch (e: Throwable) {
+                printStderr("Failure while running benchmark $name: $e\n")
```

Comment:
It's also to add here BenchmarkResult with FAILED status.

Reply:
My thought is that having an exception here is a fatal error just like `error("No matching benchmarks found")` above. I added a `try` block just to put `printStderr` in there to aid debugging if these happen.

Reply:
I think it's right to add here catching exceptions because we have one binary for several benchmarks and it's worth to run other even one fails. It's missed when it was written and after we didn't return. So I think that right way is to report failure and don't block running other benchmarks.